lib/rust/lexer/definition/src/lexer.rs

```diff
@@ -1157,3 +2320,416 @@ impl Default for BlockState {
+        if let Some(current_mut) = self.current_mut() {
+            current_mut.record_line_ending(line_ending);
+            debug!(self.logger,"Push Line Ending: {line_ending:?}");
+        }
+    }
+
+    /// Consume the last seen line ending.
+    pub fn pop_line_ending(&mut self) -> token::LineEnding {
+        if let Some(current_mut) = self.current_mut() {
+            let ending = current_mut.consume_line_ending();
+            debug!(self.logger,"Pop Line Ending: {ending:?}");
+            ending
+        } else { token::LineEnding::None }
+    }
+
+    /// Append `token` to the currently-active line in the current literal.
+    pub fn append_segment(&mut self, token:Token) {
+        debug!(self.logger,"Append Token to Current Line: {&token:?}");
+        if let Some(current_mut) = self.current_mut() {
+            current_mut.append_segment_to_line(token);
+        }
+    }
+
+    /// Consume the most-recently added token from the line.
+    pub fn consume_segment(&mut self) -> Option<Token> {
+        let result = self.current_mut().map(|t| t.consume_segment_from_line()).flatten();
+        debug!(self.logger,"Consume Segment: {result:?}");
+        result
+    }
+
+    /// Append a line to the current text literal.
+    pub fn append_line(&mut self, line:Token) {
+        debug!(self.logger,"Append Line to Current Literal: {&line:?}");
+        self.current_mut().for_each(|t| {
+            t.append_line(line);
+        });
+    }
+
+    /// Append an empty line to the current text literal with `offset` leading spaces.
+    pub fn append_empty_line(&mut self, offset:usize) {
+        let line_ending = self.pop_line_ending();
+        self.current_mut().iter_mut().for_each(|t| {
+            t.append_empty_line(offset,line_ending);
+        });
+        debug!(self.logger,"Append Empty Line: Line Ending = {line_ending:?}");
+    }
+
+    /// Submit the current line into the block.
+    pub fn submit_current_line(&mut self) {
+        self.current_mut().for_each(|line| line.submit_current_line());
+    }
+
+    /// Begin a new text literal.
+    pub fn begin_literal(&mut self) -> &mut TextState {
+        debug!(self.logger,"Begin Text Literal");
+        self.text_stack.push_and_get_mut(default())
+    }
+
+    /// End the current text literal.
+    pub fn end_literal(&mut self) -> Option<TextState> {
+        debug!(self.logger,"End Text Literal");
+        self.text_stack.pop()
+    }
+
+    /// End the current text literal.
+    ///
+    /// # Panics
+    /// Panics if there is no literal to end.
+    pub fn unsafe_end_literal(&mut self) -> TextState {
+        self.end_literal().unwrap()
+    }
+
+    /// Check if the lexer is currently in a nested text literal.
+    pub fn is_in_nested_text(&self) -> bool{
+        self.text_stack.len() > 1
+    }
+}
+
+
+
+// =================
+// === TextState ===
+// =================
+
+/// The state for lexing a single text literal.
```

Comment:
Please add more extensive docs to the structures. Describe what is inside, what is the purpose of this structure etc, so someone new to lexer would understand the struct architecture.